File: convert-litematic.py
#!/usr/bin/env python3
"""
Litematic to Blueprint Converter
Created: 2024-12-19
Converts .litematic files to agent.js blueprint format
"""

import sys
import json
import os

try:
    from litemapy import Schematic, Region, BlockState
except ImportError:
    print("Error: litemapy not found. Installing...")
    print("Please run: pip install litemapy")
    sys.exit(1)

def convert_litematic_to_blueprint(file_path):
    """Convert litematic file to blueprint format"""
    try:
        print(f"[CONVERT] Reading litematic file: {file_path}")
        
        # Load schematic
        schematic = Schematic.load(file_path)
        
        print(f"[CONVERT] Schematic name: {schematic.name}")
        print(f"[CONVERT] Number of regions: {len(schematic.regions)}")
        
        # Get the first region (or iterate through all)
        region_name = list(schematic.regions.keys())[0]
        region = schematic.regions[region_name]
        
        print(f"[CONVERT] Processing region: {region_name}")
        print(f"[CONVERT] Region size: {region.width} x {region.height} x {region.length}")
        print(f"[CONVERT] Region position: {region.x} x {region.y} x {region.z}")
        
        # Debug: Check region attributes
        print(f"[CONVERT] Region attributes: {dir(region)}")
        
        # Try to get blocks using different methods
        try:
            # Method 1: Check if there's a blocks attribute
            if hasattr(region, 'blocks'):
                print(f"[CONVERT] Found 'blocks' attribute, type: {type(region.blocks)}")
                print(f"[CONVERT] Number of blocks: {len(region.blocks) if hasattr(region.blocks, '__len__') else 'N/A'}")
        except Exception as e:
            print(f"[CONVERT] Error checking blocks attribute: {e}")
        
        # Get minimum coordinates to normalize to origin
        min_x = region.x
        min_y = region.y
        min_z = region.z
        
        # Collect all blocks
        blueprint = []
        block_count = 0
        
        print("[CONVERT] Extracting blocks...")
        
        # Use block_positions() method to iterate through all blocks
        try:
            # Get all block positions
            positions = list(region.block_positions())
            print(f"[CONVERT] Found {len(positions)} block positions")
            
            # Get minimum coordinates to normalize to origin
            if positions:
                min_x = min(pos[0] for pos in positions)
                min_y = min(pos[1] for pos in positions)
                min_z = min(pos[2] for pos in positions)
            else:
                min_x = region.x
                min_y = region.y
                min_z = region.z
            
            # Iterate through all block positions
            for x, y, z in positions:
                try:
                    # Use array-style syntax: region[x, y, z]
                    block = region[x, y, z]
                    
                    # Skip air blocks or None
                    if block is None:
                        continue
                    
                    block_id = str(block)
                    if block_id == "minecraft:air" or block_id == "air":
                        continue
                    
                    # Calculate relative position (normalize to origin)
                    rel_x = x - min_x
                    rel_y = y - min_y
                    rel_z = z - min_z
                    
                    # Remove 'minecraft:' prefix
                    block_name = block_id.replace("minecraft:", "")
                    
                    # Remove block states (everything in brackets like [snowy=false])
                    # Keep only the base block name
                    if '[' in block_name:
                        block_name = block_name.split('[')[0]
                    
                    blueprint.append({
                        'block': block_name,
                        'pos': {'x': rel_x, 'y': rel_y, 'z': rel_z}
                    })
                    block_count += 1
                except Exception as e:
                    # Debug: print first few errors
                    if block_count < 5:
                        print(f"[CONVERT] Warning at ({x}, {y}, {z}): {e}")
        except Exception as e:
            print(f"[CONVERT] Error using block_positions(): {e}")
            print("[CONVERT] Trying alternative method...")
            
            # Fallback: try allblockpos
            try:
                positions = list(region.allblockpos())
                print(f"[CONVERT] Found {len(positions)} block positions (allblockpos)")
                
                if positions:
                    min_x = min(pos[0] for pos in positions)
                    min_y = min(pos[1] for pos in positions)
                    min_z = min(pos[2] for pos in positions)
                else:
                    min_x = region.x
                    min_y = region.y
                    min_z = region.z
                
                for x, y, z in positions:
                    try:
                        block = region[x, y, z]
                        if block is None:
                            continue
                        
                        block_id = str(block)
                        if block_id == "minecraft:air" or block_id == "air":
                            continue
                        
                        rel_x = x - min_x
                        rel_y = y - min_y
                        rel_z = z - min_z
                        
                        block_name = block_id.replace("minecraft:", "")
                        
                        blueprint.append({
                            'block': block_name,
                            'pos': {'x': rel_x, 'y': rel_y, 'z': rel_z}
                        })
                        block_count += 1
                    except Exception as e:
                        if block_count < 5:
                            print(f"[CONVERT] Warning at ({x}, {y}, {z}): {e}")
            except Exception as e2:
                print(f"[CONVERT] Error using allblockpos(): {e2}")
        
        print(f"[CONVERT] Generated {block_count} non-air blocks")
        
        # Sort by Y, then Z, then X for better readability
        blueprint.sort(key=lambda b: (b['pos']['y'], b['pos']['z'], b['pos']['x']))
        
        # Output blueprint code
        print('\n// ==========================================')
        print('// Generated Blueprint Code')
        print('// ==========================================\n')
        print('const blueprint = [')
        
        for i, entry in enumerate(blueprint):
            comma = ',' if i < len(blueprint) - 1 else ''
            print(f"    {{ block: '{entry['block']}', pos: new Vec3({entry['pos']['x']}, {entry['pos']['y']}, {entry['pos']['z']}) }}{comma}")
        
        print('];\n')
        print('// ==========================================')
        print(f'// Total blocks: {block_count}')
        print('// ==========================================\n')
        
        # Also save to JavaScript format file
        output_file = 'blueprint-output.txt'
        with open(output_file, 'w', encoding='utf-8') as f:
            f.write('const blueprint = [\n')
            for i, entry in enumerate(blueprint):
                comma = ',' if i < len(blueprint) - 1 else ''
                f.write(f"    {{ block: '{entry['block']}', pos: new Vec3({entry['pos']['x']}, {entry['pos']['y']}, {entry['pos']['z']}) }}{comma}\n")
            f.write('];\n')
        
        print(f"[CONVERT] Blueprint code saved to: {output_file}")
        
        # Also save to JSON format (for agent.js to load)
        json_file = 'blueprint.json'
        blueprint_data = {
            'name': schematic.name if hasattr(schematic, 'name') else 'Blueprint',
            'source': litematic_file,
            'total_blocks': len(blueprint),
            'blocks': blueprint
        }
        
        import json
        with open(json_file, 'w', encoding='utf-8') as f:
            json.dump(blueprint_data, f, indent=2)
        
        print(f"[CONVERT] Blueprint JSON saved to: {json_file}")
        
        return blueprint
        
    except Exception as e:
        print(f"[CONVERT] Error converting litematic file: {e}")
        import traceback
        traceback.print_exc()
        return None

if __name__ == "__main__":
    litematic_file = sys.argv[1] if len(sys.argv) > 1 else '17353.litematic'
    
    if not os.path.exists(litematic_file):
        print(f"[CONVERT] File not found: {litematic_file}")
        print("[CONVERT] Usage: python convert-litematic.py <path-to-litematic-file>")
        sys.exit(1)
    
    convert_litematic_to_blueprint(litematic_file)

File: convert-litematic.js
// Litematic to Blueprint Converter
// Created: 2024-12-19
// Converts .litematic files to agent.js blueprint format

const fs = require('fs');
const path = require('path');

// Try to use prismarine-nbt for NBT parsing (litematic files are NBT-compressed)
let nbt;
try {
    nbt = require('prismarine-nbt');
} catch (e) {
    console.error('Error: prismarine-nbt not found. Installing...');
    console.error('Please run: npm install prismarine-nbt');
    process.exit(1);
}

async function convertLitematicToBlueprint(filePath) {
    try {
        console.log(`[CONVERT] Reading litematic file: ${filePath}`);
        const buffer = fs.readFileSync(filePath);
        
        // Parse NBT data
        console.log('[CONVERT] Parsing NBT data...');
        const parsed = await nbt.parse(buffer);
        const nbtData = nbt.simplify(parsed);
        
        console.log('[CONVERT] NBT structure:', JSON.stringify(Object.keys(nbtData), null, 2));
        
        // Litematic file structure:
        // - Metadata
        // - Regions (contains the actual block data)
        if (!nbtData.Metadata || !nbtData.Regions) {
            console.error('[CONVERT] Invalid litematic file structure');
            console.error('[CONVERT] Available keys:', Object.keys(nbtData));
            return;
        }
        
        const regions = nbtData.Regions;
        const regionNames = Object.keys(regions);
        
        if (regionNames.length === 0) {
            console.error('[CONVERT] No regions found in litematic file');
            return;
        }
        
        console.log(`[CONVERT] Found ${regionNames.length} region(s): ${regionNames.join(', ')}`);
        
        // Process the first region (or all regions)
        const regionName = regionNames[0];
        const region = regions[regionName];
        
        console.log(`[CONVERT] Processing region: ${regionName}`);
        console.log(`[CONVERT] Region size: ${JSON.stringify(region.Size)}`);
        console.log(`[CONVERT] Region position: ${JSON.stringify(region.Position)}`);
        
        // Get block states and palette
        const blockStates = region.BlockStates;
        const palette = region.BlockStatePalette || [];
        
        if (!blockStates || !palette) {
            console.error('[CONVERT] Missing block states or palette');
            return;
        }
        
        console.log(`[CONVERT] Palette size: ${palette.length} unique blocks`);
        console.log(`[CONVERT] Block states array length: ${blockStates.length}`);
        
        // Calculate dimensions
        const size = region.Size;
        const width = size[0] || size.x || 1;
        const height = size[1] || size.y || 1;
        const length = size[2] || size.z || 1;
        
        console.log(`[CONVERT] Dimensions: ${width} x ${height} x ${length}`);
        
        // Get position offset
        const position = region.Position || [0, 0, 0];
        const offsetX = position[0] || position.x || 0;
        const offsetY = position[1] || position.y || 0;
        const offsetZ = position[2] || position.z || 0;
        
        console.log(`[CONVERT] Position offset: (${offsetX}, ${offsetY}, ${offsetZ})`);
        
        // Calculate bits per block (for palette index)
        const bitsPerBlock = Math.max(2, Math.ceil(Math.log2(palette.length)));
        console.log(`[CONVERT] Bits per block: ${bitsPerBlock}`);
        
        // Generate blueprint array
        const blueprint = [];
        let blockCount = 0;
        
        // Litematic uses a compacted long array format
        // We need to decode the block states array
        for (let y = 0; y < height; y++) {
            for (let z = 0; z < length; z++) {
                for (let x = 0; x < width; x++) {
                    // Calculate index in the block states array
                    const index = (y * length * width) + (z * width) + x;
                    
                    // Get palette index from block states
                    // This is a simplified approach - actual decoding may be more complex
                    let paletteIndex = 0;
                    
                    if (Array.isArray(blockStates)) {
                        // If blockStates is an array of numbers
                        if (index < blockStates.length) {
                            paletteIndex = blockStates[index];
                        }
                    } else if (typeof blockStates === 'object' && blockStates.data) {
                        // If blockStates has a data array (compacted format)
                        const dataArray = blockStates.data;
                        const longIndex = Math.floor((index * bitsPerBlock) / 64);
                        const bitOffset = (index * bitsPerBlock) % 64;
                        
                        if (longIndex < dataArray.length) {
                            const longValue = BigInt(dataArray[longIndex]);
                            const mask = (BigInt(1) << BigInt(bitsPerBlock)) - BigInt(1);
                            paletteIndex = Number((longValue >> BigInt(bitOffset)) & mask);
                        }
                    }
                    
                    // Get block from palette
                    if (paletteIndex >= 0 && paletteIndex < palette.length) {
                        const blockData = palette[paletteIndex];
                        
                        // Extract block name
                        let blockName = 'air';
                        if (typeof blockData === 'string') {
                            blockName = blockData;
                        } else if (blockData && blockData.Name) {
                            blockName = blockData.Name;
                        } else if (blockData && typeof blockData === 'object') {
                            // Try to find name property
                            blockName = blockData.name || blockData.block || 'air';
                        }
                        
                        // Skip air blocks
                        if (blockName && blockName !== 'minecraft:air' && blockName !== 'air') {
                            // Remove 'minecraft:' prefix if present
                            const shortName = blockName.replace(/^minecraft:/, '');
                            
                            // Calculate relative position (subtract offset to center at origin)
                            const relX = x - offsetX;
                            const relY = y - offsetY;
                            const relZ = z - offsetZ;
                            
                            blueprint.push({
                                block: shortName,
                                pos: { x: relX, y: relY, z: relZ }
                            });
                            blockCount++;
                        }
                    }
                }
            }
        }
        
        console.log(`[CONVERT] Generated ${blockCount} non-air blocks`);
        
        // Output blueprint code
        console.log('\n// ==========================================');
        console.log('// Generated Blueprint Code');
        console.log('// ==========================================\n');
        console.log('const blueprint = [');
        
        blueprint.forEach((entry, index) => {
            const comma = index < blueprint.length - 1 ? ',' : '';
            console.log(`    { block: '${entry.block}', pos: new Vec3(${entry.pos.x}, ${entry.pos.y}, ${entry.pos.z}) }${comma}`);
        });
        
        console.log('];\n');
        console.log('// ==========================================');
        console.log(`// Total blocks: ${blockCount}`);
        console.log('// ==========================================\n');
        
        // Also save to file
        const outputFile = path.join(__dirname, 'blueprint-output.txt');
        let output = 'const blueprint = [\n';
        blueprint.forEach((entry, index) => {
            const comma = index < blueprint.length - 1 ? ',' : '';
            output += `    { block: '${entry.block}', pos: new Vec3(${entry.pos.x}, ${entry.pos.y}, ${entry.pos.z}) }${comma}\n`;
        });
        output += '];\n';
        
        fs.writeFileSync(outputFile, output);
        console.log(`[CONVERT] Blueprint code saved to: ${outputFile}`);
        
    } catch (error) {
        console.error('[CONVERT] Error converting litematic file:', error);
        console.error('[CONVERT] Stack:', error.stack);
        
        // Try alternative parsing method
        console.log('\n[CONVERT] Attempting alternative parsing method...');
        tryAlternativeParsing(filePath);
    }
}

async function tryAlternativeParsing(filePath) {
    try {
        // Try reading as raw buffer and looking for patterns
        const buffer = fs.readFileSync(filePath);
        console.log(`[CONVERT] File size: ${buffer.length} bytes`);
        
        // This is a fallback - litematic files are complex NBT structures
        console.log('[CONVERT] Litematic files require specialized parsing.');
        console.log('[CONVERT] Please consider using Litematica mod to export as .schematic or .nbt format.');
        console.log('[CONVERT] Or use an online converter tool.');
    } catch (error) {
        console.error('[CONVERT] Alternative parsing also failed:', error.message);
    }
}

// Main execution
const litematicFile = process.argv[2] || '17353.litematic';

if (!fs.existsSync(litematicFile)) {
    console.error(`[CONVERT] File not found: ${litematicFile}`);
    console.error('[CONVERT] Usage: node convert-litematic.js <path-to-litematic-file>');
    process.exit(1);
}

convertLitematicToBlueprint(litematicFile).catch(error => {
    console.error('[CONVERT] Fatal error:', error);
    process.exit(1);
});

